cowbird/handlers/impl/filesystem.py

```diff
@@ -1,79 +1,449 @@
+                                       is_executable=True)
+
+            LOGGER.debug("Creating hardlink from file `%s` to the path `%s`", src_path, hardlink_path)
+            try:
+                os.link(src_path, hardlink_path)
+            except Exception as exc:
+                LOGGER.warning("Failed to create hardlink `%s` : %s", hardlink_path, exc)
+        else:
+            LOGGER.debug("Access to the wps output file `%s` is not allowed for the user. No hardlink created.",
+                         src_path)
+
+    def _create_wpsoutputs_hardlink(self, src_path: str, overwrite: bool = False,
+                                    process_user_files: bool = True, process_public_files: bool = True) -> None:
+        regex_match = re.search(self.wps_outputs_user_data_regex, src_path)
+        access_allowed = True
+        if regex_match:  # user files
+            if not process_user_files:
+                return
+
+            # User workspace directories require write permissions to allow file modifications via JupyterLab for
+            # files with write permissions
+            is_parent_dir_writable = True
+
+            magpie_handler = HandlerFactory().get_handler("Magpie")
+            user_name = magpie_handler.get_user_name_from_user_id(int(regex_match.group(2)))
+            hardlink_path = self.get_user_hardlink(src_path=src_path,
+                                                   bird_name=regex_match.group(1),
+                                                   user_name=user_name,
+                                                   subpath=regex_match.group(3))
+            api_services = magpie_handler.get_services_by_type(ServiceAPI.service_type)
+            if SECURE_DATA_PROXY_NAME not in api_services:
+                LOGGER.debug("`%s` service not found. Considering user wpsoutputs data as accessible by default.",
+                             SECURE_DATA_PROXY_NAME)
+                apply_new_path_permissions(src_path, True, True, False)
+            else:  # get access and apply permissions if the secure-data-proxy exists
+                access_allowed = self.update_secure_data_proxy_path_perms(src_path, user_name)
+        else:  # public files
+            if not process_public_files:
+                return
+            is_parent_dir_writable = False  # public files are read-only
+            hardlink_path = self._get_public_hardlink(src_path)
+
+        if os.path.exists(hardlink_path):
+            if not overwrite and access_allowed:
+                # Hardlink already exists, nothing to do.
+                return
+            # Delete the existing file at the destination path to reset the hardlink path with the expected source.
+            LOGGER.warning("Removing existing hardlink destination path at `%s` to generate hardlink for the newly "
+                           "created file.", hardlink_path)
+            os.remove(hardlink_path)
```

Comment:
So, the way I saw the design, only the source wps outputs directory is monitored (e.g.: `/data/wps_outputs/weaver`), aka where the WPS processes write their outputs.

This source directory is synchronized with the 2 destination directories (public and user).

I was considering the destination directories to be used to give access to the WPS outputs, and not necessarily as working directories where the user could add new files manually. That's why I am deleting the file if it exists regardless, considering that the 'truth' is always found in the source directory. So, whatever if there is an existing file already in the destination (being a hardlink or an unrelated file), it would delete it, to add the true source file that should be found there.

I could always have a more forgiving behavior. Displaying a warning in the logs if it's an unrelated file and not deleting it. I can probably detect if it is a hardlink by checking the number of links on the file and also comparing the inode, but if we choose this behavior, we will never delete an existing file anyway.
A disavantage with this design, is that we won't have a guarantee that the files will be valid in the destination folders. They might be unrelated files, or actual WPS output files, but the user won't be able to know, even if we do a resync.

Reply:
> That's why I am deleting the file if it exists regardless, considering that the 'truth' is always found in the source directory.

I agree with this. Users should not have the possibility to add more files to an already executed WPS process outputs.
The best would be to actually block them with read-only directories/files, but if this cannot be achieved, the unidirectional sync from WPS outputs as "truth" is good enough for now.

Given that the user that could write additional files within its user workspace would not see the log output, I don't think it would add any value. It's good to keep it as you currently implemented it.

Reply:
The thing is the `secure-data-proxy` is adding the option of having `read` and/or `write` permissions on the user WPS output files. If we want to support `write` permissions on those files, I tested it on JupyterLab and I need to also add `write` permissions to the parent folder too, so that the file can be modified by the user via the JupyterLab interface. I didn't work if I added only the `write` permissions on the file only. Note that those folders found in the user workspace are not linked to those found in the WPS outputs source folder, so the folders don't have to use corresponding permissions with those found from the source.

So, currently, by default, all user hardlinks are `read`/`write` if the service `secure-data-proxy` doesn't exist, or, if it exists,  the permissions are resolved according to the `secure-data-proxy`'s permissions. For now, I was making the parent directory of a user hardlink `read`/`write` in every case to allow possible `write` permissions on any of their potential children files and to simplify the process, considering that the children files could have a mix of different permissions on them.

It is possible to make the user hardlinks and parent folders read-only (like the public hardlinks), but I'm not sure if it is what we want?

So I see these options :
- Keep the current design, which would make it possible for the users to create files in user WPS folder. These created files, which would be unrelated to the source WPS folder, could disappear without warning to the user if we resync the handler. In this case, we could probably add a mention about this in the eventual README (see [DAC-545](https://crim-ca.atlassian.net/browse/DAC-545)?)
- Ignore the `secure-data-proxy` entirely and making everything read-only, like the public WPS hardlinks
- Keep checking for the `secure-data-proxy`, but only use read permissions on the folders and hardlinks, ignoring the `write` permissions of the `secure-data-proxy`. This would prevent the user from creating unrelated files in its user WPS output folder.

Reply:
I think it is a good idea to have some details in a README about automatic removal of the files due to how sync could behave, such that the user can be informed of why his files could disappear, and why it is not recommended to add any more files in there. 

I'm not too worried about the `write` on the directory, since the unidirectional sync will not propagate back to the source WPS outputs. However, doesn't `write` on the hardlink allow the user to modify the original file? For example, doing something like `cat "random" >> "<hardlink/file>"` could modify the source WPS output? If this is the case, then it is something that should definitely be prevented with `read` only access using the permissions of `secure-data-proxy`.

Reply:
Yes, indeed. The `write` on the hardlink does modify the original source file, and I was aware of that. I assumed it was okay, since it is a user-specific file, and the `secure-data-proxy` is allowing specification of `write` permissions on the file. 
I can adjust it if we want to change this behavior. Should I ignore `write` permissions on `secure-data-proxy` on Magpie though? It seems it might lead to confusion if a `write` permissions is added but the file stays read-only.

But, yes I can change it. I could only check the `read` permissions on `secure-data-proxy` to adjust the file permissions, and I could assign `read-only` on the hardlinks by default if the `secure-data-proxy` service doesn't exist, ignoring all `write` permissions.

Let me know if we should go ahead with this design, or if you have any idea about how to avoid confusion with the `write` permissions on Magpie (except to document it in .rst files).

Reply:
Yes. It should be modified, but don't ignore the permission in Cowbird code. If `write` is explicitly set, then it would be valid to let override of the file by the user. Cowbird would sync the permission normally. This would make sense in the case of `secure-data-proxy` for other directories than WPS outputs. However, WPS outputs themselves should never have `write`, since they are produced by some processes and their result should remain constant. If `write` was set somewhere for WPS output directories (whether for `{user}` or `public`), that would be an error (ie: bad Cowbird *configuration* or some improperly defined Magpie hook).